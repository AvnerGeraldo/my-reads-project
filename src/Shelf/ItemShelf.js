import React from 'react'
import PropTypes from 'prop-types'

//Assets
import './ItemShelf.css'

const ItemShelf = ({ bookData, changeBookShelf }) => {
    //Vars
    const bookTitle = bookData.title !== undefined ? bookData.title : ''
    const backgroundImage = bookData.imageLinks !== undefined && bookData.imageLinks.thumbnail !== undefined ? `url(${bookData.imageLinks.thumbnail}}` : null
    const valueSelected = (bookData.shelf === null || bookData.shelf === undefined ? 'none' : bookData.shelf)
    const listAuthors = bookData.authors !== undefined ? bookData.authors : []

    //Style
    const styleBookCover = {
        backgroundImage,
        width: 128,
        height: 193
    }    

    return (
        <div className="book">
            <div className="book-top">
            <div className="book-cover" style={styleBookCover}></div>
                <div className="book-shelf-changer">
                    <select value={valueSelected} onChange={(e) => changeBookShelf(bookData, e.target.value)}>
                        <option value="move" disabled>Move to...</option>
                        <option value="currentlyReading">Currently Reading</option>
                        <option value="wantToRead">Want to Read</option>
                        <option value="read">Read</option>
                        <option value="none">None</option>
                    </select>
                </div>
            </div>
            <div className="book-title">{bookTitle}</div>
            <div className="book-authors">{listAuthors.join(' , ')}</div>
        </div>
    )
}

const { object, func } = PropTypes
ItemShelf.propTypes = {
    bookData: object.isRequired,
    changeBookShelf: func.isRequired
}

export default ItemShelf