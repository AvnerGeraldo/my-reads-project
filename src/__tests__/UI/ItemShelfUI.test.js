import React from 'react'
import { shallow } from 'enzyme'

//Component
import ItemShelf from '../../Shelf/ItemShelf'

const props = {
    bookData: {
        title: 'book title',
        imageLinks: {
            thumbnail: 'image-link'
        },
        shelf: 'wantToRead',
        authors: ['author1', 'author2']
    }, 
    changeBookShelf: jest.fn()
}

describe('[Component] ItemShelf', () => {
    it('render item book correctly', () => {
        const wrapper = shallow(<ItemShelf {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('initialize select with \'none\' value when \'bookData.shelf\' is null or not exists', () => {
        const newProps = { ...props, bookData: { ...props.bookData, shelf: null }}
        const wrapper = shallow(<ItemShelf {...newProps} />)

        expect(wrapper.find('select').props().value).toBe('none')
        expect(wrapper.find('select').props().value).not.toBeNull()
        expect(wrapper.find('select').props().value).not.toBeUndefined()
    })

    it('verified if book title is rendered', () => {
        const wrapper = shallow(<ItemShelf {...props} />)

        expect(wrapper.find('.book-title').text()).toBeDefined()
        expect(wrapper.find('.book-title').text()).not.toBeNull()
        expect(wrapper.find('.book-title').text()).toEqual(props.bookData.title)
    })

    it('verified if book authors is rendered', () => {
        const wrapper = shallow(<ItemShelf {...props} />)

        expect(wrapper.find('.book-authors').text()).toBeDefined()
        expect(wrapper.find('.book-authors').text()).not.toBeNull()
        expect(wrapper.find('.book-authors').text()).toEqual(props.bookData.authors.join(' , '))
    })
})