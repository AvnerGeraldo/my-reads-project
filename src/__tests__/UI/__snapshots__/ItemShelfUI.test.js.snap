// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] ItemShelf render item book correctly 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(image-link}",
          "height": 193,
          "width": 128,
        }
      }
    />
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value="wantToRead"
      >
        <option
          disabled={true}
          value="move"
        >
          Move to...
        </option>
        <option
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          value="read"
        >
          Read
        </option>
        <option
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  >
    book title
  </div>
  <div
    className="book-authors"
  >
    author1 , author2
  </div>
</div>
`;
